### PR TITLE
pvc storage request warning for fractional byte value

### PR DESCRIPTION
**Pull request by pacoxu**
#### What type of PR is this?
/kind bug

#### Which issue(s) this PR fixes:
xref https://github.com/kubernetes/kubernetes/issues/94313

In https://github.com/kubernetes/kubernetes/issues/94626
- Add warnings for known bad values
- - PVC, PV, PVC template in statefulset and inline ephemeral volume
- - - fractional byte storage requests  

#### Special notes for your reviewer:
I followed #101688

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

### CODE REVIEW

**Comment by liggitt** [comment]
/hold

should this also be run on the PVC template in statefulset and in pod ephemeral volume source?

File: pkg/api/persistentvolumeclaim/util.go [hunk 1]
```diff
@@ -151,3 +154,18 @@ func allocatedResourcesInUse(oldPVC *core.PersistentVolumeClaim) bool {
...
+		return nil
+	}
+	storageValue := pv.Spec.Resources.Requests[core.ResourceStorage]
+	var warnings []string
+	if storageValue.MilliValue()%int64(1000) != int64(0) {
+		warnings = append(warnings, fmt.Sprintf(
+			"%s: fractional byte value %q is invalid, must be an integer",
+			field.NewPath("spec").Child("resources").Child("requests").Key(core.ResourceStorage.String()),
+			storageValue.String()))
+	}
```
Comment (liggitt) [inline pkg/api/persistentvolumeclaim/util.go:169]: wondering if this bit should be factored into a GetWarningsForPersistentVolumeClaimSpec function we can call when getting warnings for statefulsets and podspecs

Comment (pacoxu) [reply pkg/api/persistentvolumeclaim/util.go:169]: Good point. I thought this is not used by other functions and merged the two functions together. If there are other usages, it should be refactored.

File: pkg/api/pod/warnings.go [hunk 2]
```diff
@@ -151,6 +152,9 @@ func warningsForPodSpecAndMeta(fieldPath *field.Path, podSpec *api.PodSpec, meta
 		if v.Glusterfs != nil {
 			warnings = append(warnings, fmt.Sprintf("%s: deprecated in v1.25, this feature will be removed soon after in a subsequent release", fieldPath.Child("spec", "volumes").Index(i).Child("glusterfs")))
 		}
+		if v.Ephemeral != nil {
```
Comment (liggitt) [inline pkg/api/pod/warnings.go:155]: make sure VolumeClaimTemplate is not null as well

Comment (pacoxu) [reply pkg/api/pod/warnings.go]: Done.

And fix the import verify failure.

File: pkg/api/persistentvolumeclaim/util.go [hunk 3]
```diff
@@ -151,3 +154,23 @@ func allocatedResourcesInUse(oldPVC *core.PersistentVolumeClaim) bool {
...
+func GetWarningsForPersistentVolumeClaim(pv *core.PersistentVolumeClaim) []string {
+	if pv == nil {
+		return nil
+	}
+
+	return GetWarningsForPersistentVolumeClaimSpec(field.NewPath("spec"), pv.Spec)
+}
+
+func GetWarningsForPersistentVolumeClaimSpec(fieldPath *field.Path, pvSpec core.PersistentVolumeClaimSpec) []string {
+	storageValue := pvSpec.Resources.Requests[core.ResourceStorage]
```
Comment (liggitt) [inline pkg/api/persistentvolumeclaim/util.go:167]: do we also need to look at `pvSpec.Resources.Limits`?

Comment (pacoxu) [reply pkg/api/persistentvolumeclaim/util.go]: Added.

https://github.com/kubernetes/kubernetes/blob/9c4905e7af0d070d139c06638ab9868fd3521883/plugin/pkg/admission/limitranger/admission.go#L484
At first, I saw this in comments and not add the limit checking.

**Comment by liggitt** [comment]
/lgtm
/approve
/hold cancel

thanks for adding this :)

**Pull request merged**